modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/utils/fieldSupport.js

```diff
@@ -450,6 +450,51 @@ export function updateEditorConfigInstanceId(editorConfig, instanceId) {
 	return updatedEditorConfig;
 }
 
+export function updateInputMaskProperties(editingLanguageId, field) {
```

Comment:
Why do we need to map the fields two times? Can't we use one `mapFields` call only?

Reply:
because I'm not sure whether `visitor.mapFields` will return the `inputMaskFormat` and `numericInputMask` first so I can get the localized value and then update the `validation` and `predefinedValue` fields.  If the `predefinedValue` field comes first, I will set its `inputMaskFormat` value to empty. What do you think?

Reply:
It makes sense. Do you think we could at least save the index of the `predefinedValue` and `validation`, so we could access them directly after the first `mapFields`?
Also, just another thing I saw is that you don't need to assign the field to a "new field" variable. The `mapFields` function already merges properties, by default. So, instead of:
```
newField = {
	...newField,
	...numericInputMask,
	inputMaskFormat,
};
```

You could just return the properties to be updated like so:
```
return {
	...numericInputMask,
	inputMaskFormat,
};
```
Does it make sense?

Reply:
Got it! I'll try your suggestions

Reply:
>  Do you think we could at least save the indexes of the predefinedValue and validation, or perhaps the field itself, so we could access them directly after the first mapFields?

In this case, I would have to save the `columnIndex, pageIndex, rowIndex and fieldIndex` and then do something like:

```
newField.settingsContext.pages[pageIndex].rows[rowIndex].columns[columnIndex].fields[fieldIndex] = {
     ...newField.settingsContext.pages[pageIndex].rows[rowIndex].columns[columnIndex].fields[fieldIndex],
     ...numericInputMask,
     inputMaskFormat,
}
```

Do you think it's worth it?

Reply:
If I just save the field, I still have to do `newField.settingsContext.pages[pageIndex].rows[rowIndex].columns[columnIndex].fields[fieldIndex] = {`

Reply:
I was thinking of something like this:
```
export function updateInputMaskProperties(editingLanguageId, field) {
	let inputMaskFormat = '';
	let numericInputMask = {};
	let predefinedValueField;
	let validationField;

	const visitor = new PagesVisitor(field.settingsContext.pages);

	visitor.mapFields((field) => {
		if (field.fieldName === 'inputMaskFormat') {
			inputMaskFormat = field.localizedValue?.[editingLanguageId];
		}
		else if (field.fieldName === 'numericInputMask') {
			numericInputMask = field.localizedValue?.[editingLanguageId];

			if (typeof numericInputMask === 'string') {
				numericInputMask = JSON.parse(numericInputMask);
			}
		}
		else if (field.fieldName === 'predefinedValue') {
			predefinedValueField = field;
		}
		else if (field.fieldName === 'validation') {
			validationField = field;
		}

		return {
			...numericInputMask,
			inputMaskFormat,
		};
	});
	
	predefinedValueField.inputMaskFormat = inputMaskFormat;
	predefinedValueField.numericInputMask = numericInputMask;
	
	validationField.inputMaskFormat = inputMaskFormat;
	validationField.numericInputMask = numericInputMask;
}
```
Do you think this could work?

Reply:
I tried, but this way the value is not updated inside the `settingsContext` of the `newField`. That’s why we do:

`newField.settingsContext.pages = visitor.mapFields((field) => {` in the second `mapFields`

Reply:
@rodrigopaulino I created a new test ([here](https://github.com/liferay-forms/liferay-portal/pull/1291/commits/3cdc0277d10fc8ce6d23d0f0dd843252fe1e99fa#diff-80ee30b37d95463de2abc1e23b06844d5be145a4864d66be3294d9c951ab7e15R209)) to validate the expected behavior of `updateInputMaskProperties`.

When I try to use `let predefinedValueField; let validationField;`, as suggested, the test fails. The only way I can think of to avoid the second `mapFields` is to use `field.settingsContext.pages[pageIndex].rows [rowIndex].columns[columnIndex].fields[fieldIndex]`, which I don't like.

I also made some changes [here](https://github.com/liferay-forms/liferay-portal/pull/1291/commits/eaafef9cfb99b5797d2c46d3923d799678750557#diff-9243a95a9886707869a4d96938ccda27c36468ef0a575f1ecb1222178f63ca45R472).

Reply:
Hey, @carolmariaabb. Could you verify if the following could work for you?

```
export function updateInputMaskProperties(editingLanguageId, field) {
	let inputMaskFormat = '';
	let numericInputMask = {};
	let predefinedValueField;
	let validationField;

	const visitor = new PagesVisitor(field.settingsContext.pages);

	visitor.visitFields((setting) => {
		if (setting.fieldName === 'inputMaskFormat') {
			inputMaskFormat = setting.localizedValue?.[editingLanguageId];
		}
		else if (setting.fieldName === 'numericInputMask') {
			numericInputMask = setting.localizedValue?.[editingLanguageId];

			if (typeof numericInputMask === 'string') {
				numericInputMask = JSON.parse(numericInputMask);
			}
		}
		else if (setting.fieldName === 'predefinedValue') {
			predefinedValueField = setting;
		}
		else if (setting.fieldName === 'validation') {
			validationField = setting;
		}
	});

	field.inputMaskFormat = inputMaskFormat;
	predefinedValueField.inputMaskFormat = inputMaskFormat;
	validationField.inputMaskFormat = inputMaskFormat;

	Object.keys(numericInputMask).forEach((key) => {
		field[key] = numericInputMask[key];
		predefinedValueField[key] = numericInputMask[key];
		validationField[key] = numericInputMask[key];
	});

	return field;
}
```

To my understanding, you don't even need to return anything in this function. Just call the update function and the references will be updated.

Reply:
Oh! Forgot to mention, great job on the new test! 👏🏻

Reply:
It worked! I was trying to do

```
predefinedValueField = {
	...predefinedValueField,
	...numericInputMask,
	inputMaskFormat,
}
```
and it wasn’t working. I was running out of ideas.  

> To my understanding, you don't even need to return anything in this function.

I also removed the return. 

Thanks a lot, Paulino!

![giphy](https://user-images.githubusercontent.com/34186632/147487024-e9b3eae4-d786-4e63-85d2-a7003c5dce20.gif)

Reply:
Awesome!!